upgrade/upgrade-scripts/execute

```diff
@@ -69,92 +70,93 @@ EOF
+#    versions of each package, this would prevent packages from being
+#    upgraded without either:
 #
-apt_get install -y "delphix-platform-$platform" ||
-	die "failed to install delphix-platform"
-
+#        a) also removing the "delphix-entire" package, or
+#        b) building and installing new "delphix-entire" package that
+#           depends on the new package being installed.
 #
-# Now that we know we have a delphix-platform-<platform> packages
-# installed, go ahead and install delphix-entire-<platform>.
+#    Both options, (a) and (b), are undesireable; by not using package
+#    dependencies we avoid having to deal this problem.
 #
+
 source_version_information
-apt_get install -y "delphix-entire-$platform=$VERSION" ||
+apt_get install -y --allow-downgrades "delphix-entire-$platform=$VERSION" ||
 	die "upgrade failed; from '$INSTALLED_VERSION' to '$VERSION'"
 
+[[ -f "/usr/share/doc/delphix-entire-$platform/packages.list.gz" ]] ||
+	die "delphix-entire's packages.list.gz file is missing"
+
+apt_get install -y --allow-downgrades \
+	$(zcat /usr/share/doc/delphix-entire-$platform/packages.list.gz) ||
+	die "failed to install packages listed in packages.list.gz file"
+
 #
-# After we've successfully installed the new delphix-entire package, we
-# need to remove all packages that are no longer required. For example,
-# if the old version depended on a package, and the new version no
-# longer depends on that package, the upgrade performed above will not
-# remove this package. Thus, we need to run the "autoremove" to remove
-# these packages that are no longer required.
+# After we've successfully installed the new packages, we need to remove
+# all packages that are no longer required. For example, if the old
+# "delphix-entire" version required a package, and the new version no
+# longer requires that package, the commands run above will not remove
+# the package. Thus, we need to run "autoremove" to ensure all packages
+# that fall into this category are removed.
 #
 apt_get autoremove --purge -y || die "autoremove after upgrade failed"
 
 #
-# Finally, for all of the packages that delphix-entire depends on, we
-# verify the package is installed and its version is correct; this is
-# simply to help us be confident that upgrade behaves as we expect.
+# Finally, for all of the packages listed in the "packages.list" file,
+# we verify the package is installed and its version is correct; this
+# is simply to help us be confident that upgrade behaves as we expect.
 #
-dpkg-query -Wf '${Depends}' "delphix-entire-$platform" |
-	sed 's/, /\n/g' |
-	sed 's/^\(.*\) (\(.*\) \(.*\))$/\1 \2 \3/g' |
-	while read -r package op dependency; do
-		installed=$(dpkg-query -Wf '${Version}' "$package")
-		compare_versions "$installed" "$op" "$dependency" ||
-			die "'$package' package version incompatible;" \
-				"'$installed' '$op' '$dependency'"
+zcat "/usr/share/doc/delphix-entire-$platform/packages.list.gz" | sed 's/=/ /' |
```

Comment:
I think it would be valuable to also print all the packages that are installed that are not part of the current delphix-entire list. This should show packages pinned by the previous delphix-kernel version (given the changes I proposed above). This would be quite useful for debugging.

Reply:
Do you think this is something we should do as part of the upgrade scripts? IMO, that's just adding more output that'll likely get ignored, so I'm wondering if that'd be better left as something we do when gathering a support bundle.. what do you think?

Reply:
Yeah, I think that's a good idea to put that into the bundle instead

Reply:
OK, I'll open a bug to get that done.